File: classics/LongestIncreasingSubSequence.cpp
#include <cstdio>
#include <vector>
#include <algorithm>
using namespace std;

#define MAX 5001

vector<int> S(MAX,1);
vector<int> seq;
int n;

int LIS() {
    
	for ( int i=1; i<n; i++ ) {
		for ( int j=0;j<i; j++ ) {
			if ( seq[j] < seq[i] && S[j] >= S[i] ) {
				S[i]=S[j]+1;
			}
		}
	}
	
	int max=0;
	for ( int i=0; i<MAX; i++ ) {
		if ( S[i] > max ) {
			max=S[i];
		}
	}
	
	return max;
}

int main ( void ) {

	scanf("%d", &n);
	for ( int i=0; i<n; i++ ) {
		int ni;
		scanf("%d", &ni);
		seq.push_back(ni);
	}

	int ans = LIS();
	printf("%d\n", ans);
	return 0;
	
}

File: classics/LongestIncreasingSubSequence.py
import sys

seq=[]
best=[]
n=0

def LIS():
	for i in range(n):
		optimal=0
		for j in range(i-1, -1, -1):
			if seq[j] < seq[i] and best[j] > optimal:
				optimal=best[j]
			if j <= optimal: break
		best[i]=optimal+1
	return max(best)

allData = map( int, sys.stdin.read().split() )

n=allData[0]
seq=allData[1:]
best=[1]*n

print LIS()
